### PR TITLE
Callback code generation rework

### PR DESCRIPTION
**Pull request by FeralChild64**
# Description

I intend to change the code of one callback, so that it behaves more like the real BIOS. Unfortunately, the current code generator is rather hard to tamper with, one needs to constantly watch the offset from the start of the code and adapt it accordingly for every instruction after the inserted/deleted code.

This rework is intended to make the generator code a little bit easier to maintain. There is a potential for regressions (despite I have self-reviewed the code twice) - but it should be easy to withdraw specific parts of the rework to find out the root cause.

## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/3691 (rework related to the issue; PR does not fix the issue)

# Manual testing

Tried to run CtMouse 2.0, CtMouse 2.1, DOS Navigator, several games (Warcraft II, Settlers II, Colonization, etc.) - no problems found.

# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by johnnovak** [comment]
@FeralChild64 It's WIP I know, but you seem to be mixing huge non-functional reformats with actual code changes.

Could you separate the two? Ideally we'd have many separate commits for easy bisecting. Pure reformat commits can be big, that doesn't matter, but please don't mix in functional changes and refactors. Code refactors and changes/additions should be as many little self-contained changes as possible. The more, the better, so we can pinpoint what exactly introduced a regression.

Thanks 😄

File: src/cpu/callback.cpp [hunk 1]
```diff
@@ -154,29 +155,54 @@ const char* CALLBACK_GetDescription(callback_number_t cb_num)
...
 
-callback_number_t CALLBACK_SetupExtra(callback_number_t cb_num, Bitu type, PhysPt physAddress, bool use_cb = true)
+static uint8_t callback_setup_extra(const callback_number_t callback_number,
+                                    const uint8_t callback_type,
+                                    const PhysPt start_address,
+                                    const bool should_use_callback = true)
 {
-	if (cb_num >= CB_MAX)
+	if (callback_number >= CB_MAX) {
+		E_Exit("Unknown callback number 0x%04x", callback_number);
```
Comment (johnnovak) [inline src/cpu/callback.cpp:164]: Would it be better to just log an error and return without hard exiting? I really dislike hard exits.

File: src/cpu/callback.cpp [hunk 2]
```diff
@@ -543,17 +569,19 @@ callback_number_t CALLBACK_SetupExtra(callback_number_t cb_num, Bitu type, PhysP
 			phys_writeb(physAddress+0x1B,(uint8_t)0xC3);	//A RETN Instruction
 		return (use_cb?32:27);
 	default:
-		E_Exit("CALLBACK:Setup:Illegal type %" sBitfs(u),type);
+		E_Exit("Unknown callback type 0x%04x", callback_type);
```
Comment (johnnovak) [inline src/cpu/callback.cpp:525]: Would it be better to just log an error and return without hard exiting? I really dislike hard exits.

File: src/cpu/callback.cpp [hunk 3]
```diff
@@ -154,29 +155,54 @@ const char* CALLBACK_GetDescription(callback_number_t cb_num)
 	return CallBack_Description[cb_num].c_str();
 }
 
-callback_number_t CALLBACK_SetupExtra(callback_number_t cb_num, Bitu type, PhysPt physAddress, bool use_cb = true)
+static uint8_t callback_setup_extra(const callback_number_t callback_number,
+                                    const uint8_t callback_type,
+                                    const PhysPt start_address,
+                                    const bool should_use_callback = true)
```
Comment (johnnovak) [inline src/cpu/callback.cpp:161]: minor: just `use_callback`?

File: src/cpu/callback.cpp [hunk 4]
```diff
@@ -154,29 +155,54 @@ const char* CALLBACK_GetDescription(callback_number_t cb_num)
...
 {
-	if (cb_num >= CB_MAX)
+	if (callback_number >= CB_MAX) {
+		E_Exit("Unknown callback number 0x%04x", callback_number);
 		return 0;
-	switch (type) {
+	}
+
+	// TODO: temporary, until rework is complete
+	auto physAddress = start_address;
```
Comment (johnnovak) [inline src/cpu/callback.cpp]: `phys_address`

File: src/cpu/callback.cpp [hunk 5]
```diff
@@ -195,379 +212,315 @@ static uint8_t callback_setup_extra(const callback_number_t callback_number,
...
-			phys_writeb(physAddress+0x01,(uint8_t)0x38);	//Extra Callback instruction
-			phys_writew(physAddress + 0x02, cb_num);        // The immediate word
-			physAddress += 4;
+		if (should_use_callback) {
+			add_native_call_4(callback_number);
 		}
-		phys_writeb(physAddress+0x00,(uint8_t)0xCA);		//A RETF 8 Instruction
-		phys_writew(physAddress+0x01,(uint16_t)0x0008);
-		return (use_cb?7:3);
+		add_instruction_3(0xca, 0x08, 0x00);       // retf 8
```
Comment (johnnovak) [inline src/cpu/callback.cpp:227]: Oh man, manually returning and maintaining all these values... Why do people do such things to themselves... 😕

File: src/cpu/callback.cpp [hunk 6]
```diff
@@ -195,379 +212,315 @@ static uint8_t callback_setup_extra(const callback_number_t callback_number,
...
+		if (should_use_callback) {
+			add_native_call_4(callback_number);
 		}
-		phys_writeb(physAddress+0x01,(uint8_t)0xCF);		//An IRET Instruction
-		for (uint8_t i = 0; i <= 0x0b; ++i)
-			phys_writeb(physAddress + 0x02 + i, 0x90);
-		phys_writew(physAddress + 0x0e, (uint16_t)0xedeb); // jmp callback
-		return (use_cb ? 0x10 : 0x0c);
+		add_instruction_1(0xcf);                   // iret
+		for (uint8_t i = 0; i <= 0x0b; ++i) {
```
Comment (johnnovak) [inline src/cpu/callback.cpp:393]: minor `auto i`

**Review by johnnovak — APPROVED** [review]
Looks like a major maintainability improvement @FeralChild64 , very nice 🥇 I have only reviewed the generic principle behind this (which is good), and spot checked a few of the conversions—all look good. I guess if there's a typo somewhere, we'll find it out sooner or later, and it should be easy to verify the correctness if there's a regression.

File: src/cpu/callback.cpp (hunk 5 above)
Comment (FeralChild64) [reply src/cpu/callback.cpp:227]: I can't understand this neither. True, they didn't have lambdas back then - but passing a reference to current address variable would have improved this, too.

File: src/cpu/callback.cpp (hunk 6 above)
Comment (FeralChild64) [reply src/cpu/callback.cpp]: OK, changed.

File: src/cpu/callback.cpp (hunk 3 above)
Comment (FeralChild64) [reply src/cpu/callback.cpp]: OK, changed.

File: src/cpu/callback.cpp (hunk 1 above)
Comment (FeralChild64) [reply src/cpu/callback.cpp]: OK, changed - no more `E_Exit` in the reworked function.

File: src/cpu/callback.cpp (hunk 2 above)
Comment (FeralChild64) [reply src/cpu/callback.cpp]: OK, changed - no more `E_Exit` in the reworked function.

**Pull request merged**